File: cpp/data_structures/array/array.cpp
#include <iostream>
#include <climits> // For INT_MAX and INT_MIN
using namespace std;

// Function to display the elements of an array
void displayArray(const int arr[], int size) {
    cout << "Array elements: ";
    for (int i = 0; i < size; ++i) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

// Function to search for an element in an array
bool searchElement(const int arr[], int size, int target) {
    for (int i = 0; i < size; ++i) {
        if (arr[i] == target) {
            return true;
        }
    }
    return false;
}

// Function to find the maximum element in an array
int findMax(const int arr[], int size) {
    int maxElement = INT_MIN;
    for (int i = 0; i < size; ++i) {
        if (arr[i] > maxElement) {
            maxElement = arr[i];
        }
    }
    return maxElement;
}

// Function to find the minimum element in an array
int findMin(const int arr[], int size) {
    int minElement = INT_MAX;
    for (int i = 0; i < size; ++i) {
        if (arr[i] < minElement) {
            minElement = arr[i];
        }
    }
    return minElement;
}

// Function to calculate the sum of elements in an array
int calculateSum(const int arr[], int size) {
    int sum = 0;
    for (int i = 0; i < size; ++i) {
        sum += arr[i];
    }
    return sum;
}

// Function to calculate the average of elements in an array
double calculateAverage(const int arr[], int size) {
    int sum = calculateSum(arr, size);
    return static_cast<double>(sum) / size;
}

// Function to calculate the median of a sorted array
double calculateMedian(const int arr[], int size) {    
    if (size % 2 == 0) {
        // If the size of the array is even, average the middle two elements
        int middle1 = size / 2 - 1;
        int middle2 = size / 2;
        return (static_cast<double>(arr[middle1]) + arr[middle2]) / 2.0;
    } else {
        // If the size of the array is odd, the middle element is the median
        int middle = size / 2;
        return arr[middle];
    }
}

int main() {
    // Array Initialization
    int arr[] = { 1, 2, 3, 4, 5 };
    int size = sizeof(arr) / sizeof(arr[0]);

    // Array Traversal
    displayArray(arr, size);

    // Find Maximum and Minimum
    cout << "Maximum element: " << findMax(arr, size) << endl;
    cout << "Minimum element: " << findMin(arr, size) << endl;

    // Calculate Sum, Average, and Median
    cout << "Sum of elements: " << calculateSum(arr, size) << endl;
    cout << "Average of elements: " << calculateAverage(arr, size) << endl;
    cout << "Median of elements: " << calculateMedian(arr, size) << endl;

    // Search for an Element
    int target;
    cout << "Enter an element to search: ";
    cin >> target;
    if (searchElement(arr, size, target)) {
        cout << target << " found in the array.\n";
    }
    else {
        cout << target << " not found in the array.\n";
    }

    return 0;
}

File: cs/data_structures/array/array.cs
using System;

class MyArray
{
    private int[] array;

    public MyArray(int[] arr)
    {
        array = arr;
    }

    public void Display()
    {
        Console.Write("Array: ");
        foreach (int num in array)
        {
            Console.Write(num + " ");
        }
        Console.WriteLine();
    }

    public void Search(int element)
    {
        int index = Array.IndexOf(array, element);
        if (index != -1)
        {
            Console.WriteLine($"Element {element} found at index {index}");
        }
        else
        {
            Console.WriteLine($"Element {element} not found in the array");
        }
    }

    public void FindMax()
    {
        int max = array[0];
        foreach (int num in array)
        {
            if (num > max)
            {
                max = num;
            }
        }
        Console.WriteLine($"Maximum element: {max}");
    }

    public void FindMin()
    {
        int min = array[0];
        foreach (int num in array)
        {
            if (num < min)
            {
                min = num;
            }
        }
        Console.WriteLine($"Minimum element: {min}");
    }

    public void CalculateAverage()
    {
        double sum = 0;
        foreach (int num in array)
        {
            sum += num;
        }
        double average = sum / array.Length;
        Console.WriteLine($"Average of the array: {average}");
    }

    public void FindMedian()
    {
        // Note: The array must be sorted to find the median
        Array.Sort(array);

        int middle = array.Length / 2;

        if (array.Length % 2 == 0)
        {
            // Even number of elements, average the middle two
            double median = (array[middle - 1] + array[middle]) / 2.0;
            Console.WriteLine($"Median of the array: {median}");
        }
        else
        {
            // Odd number of elements, middle element is the median
            Console.WriteLine($"Median of the array: {array[middle]}");
        }
    }
}

class Program
{
    static void Main()
    {
        int[] numbers = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 };

        MyArray myArray = new MyArray(numbers);

        myArray.Display();
        myArray.Search(7);
        myArray.FindMax();
        myArray.FindMin();
        myArray.CalculateAverage();
        myArray.FindMedian();
    }
}
